## Lab: SQL Practice

> Author: Matt Brems

In this lab, you're going to have ten prompts. Each prompt will require you to do two things:
1. Write a SQL query.
2. Use the result of that query to answer the question.

The purpose of this lab is to get you to practice your SQL - an **important** skill in data science! While it is possible to get these answers using Pandas, it should be your goal to do everything only in SQL.

**Data**: The data used in this lab is the Iowa liquor database. It contains various tables of transactions. The schema for this database is here:

<img src="./images/schema.png" alt="schema" width="750"/>

In [159]:
# Import Pandas and Create_Engine

import pandas as pd
from sqlalchemy import create_engine

# Set up the engine to access the data.
engine = create_engine('postgres://analytics_student:analyticsga@analyticsga-psql.generalassemb.ly:5432/iowa_liquor_sales_database')

### Prompt 1

How many rows are there in the `stores` table of the Iowa liquor database?

In [160]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT COUNT(stores) 
FROM stores
"""

pd.read_sql_query(sql, engine)

count
0   1973

**Prompt 1 Answer**: 1973

### Prompt 2

If you sort the stores in alphabetical order by name, which store is fourth?

In [161]:
# Use a SQL query to find the answer to the above prompt.

sql = """

SELECT *
FROM stores
ORDER BY name ASC

"""

pd.read_sql_query(sql, engine)

store                                name store_status  \
0      4986  218 Fuel Express & Chubby's Liquor            A   
1      3047          3047 Cub Foods / Iowa City            I   
2      3060               3060 Cub Foods / Ames            I   
3      3061         3061 Cub Foods / Sioux City            I   
4      4845                  3rd St Convenience            A   
...     ...                                 ...          ...   
1968   3992       Xo Food And Liquor / Waterloo            I   
1969   4066                       Xpress Liquor            I   
1970   4886                      Yr Dollar Stop            A   
1971   4908                Zapf's Pronto Market            A   
1972   4874                      Z's Quickbreak            A   

                                          store_address address_info  
0     68 Monroe St\nFloyd, IA 50435\n(43.12249790500...         None  
1     855 Highway 1 West\nIowa City, IA 522460000\n(...         None  
2     3121 Grand Ave\nAmes, IA 500100000\n(42.053458...         None  
3     1732 Hamilton Blvd\nSioux City, IA 511030000\n...         None  
4     216, 3rd St Se\nCedar Rapids, IA 52401\n(41.97...         None  
...                                                 ...          ...  
1968  428 Franklin St\nWaterloo, IA 507030000\n(42.5...         None  
1969  1201 North Jefferson Ste 100\nIndianola, IA 50...         None  
1970  1500 1st Ave Ne\nCedar Rapids, IA 52402\n(41.9...         None  
1971  107 1st St Sw\nElkader, IA 52043\n(42.85350191...         None  
1972  201, E Washington St\nMount Pleasant, IA 52641...         None  

[1973 rows x 5 columns]

**Prompt 2 Answer**: 3061 Cub Foods / Sioux City

### Prompt 3

Among those sales whose category name is `IMPORTED VODKA`, if you sort by vendor in alphabetical order, what is the bottle price and number of bottles bought in the first 3 transactions?
> You should give us three sets of numbers - one for each transaction.

In [162]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT *
FROM sales
WHERE category_name = 'IMPORTED VODKA'
ORDER BY vendor
LIMIT 3
"""

pd.read_sql_query(sql, engine)

date convenience_store  store county_number     county category  \
0 2014-07-01              None   4509            85      Story  1032080   
1 2014-10-08              None   4046             4  Appanoose  1032080   
2 2014-07-30              None   2548            77       Polk  1032080   

    category_name vendor_no                     vendor    item  \
0  IMPORTED VODKA        14  Adamba Imports Int'l Inc.  904783   
1  IMPORTED VODKA        10      A Hardy / U.S.A. Ltd.   36200   
2  IMPORTED VODKA        10      A Hardy / U.S.A. Ltd.   36200   

                   description  pack  liter_size state_btl_cost btl_price  \
0                 Vesica Vodka    12         750          $5.99     $8.99   
1  Michael Godard "Mr. Smooth"     6         750         $19.00    $28.50   
2  Michael Godard "Mr. Smooth"     6         750         $19.00    $28.50   

   bottle_qty   total  
0          12  107.88  
1           1   28.50  
2           6  171.00

**Prompt 3 Answer**: 

In [163]:
# $8.99, 12
# $28.50, 1
# $28.50, 6

### Prompt 4

What is the total amount of liquor sold in February 2015?

In [164]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT SUM(total)
FROM sales
WHERE date between '2015-02-01' and '2015-02-28' 
"""

pd.read_sql_query(sql, engine)

sum
0  21295350.55

**Prompt 4 Answer**: $21,295,350.55

### Prompt 5

Among all transactions where `IMPORTED VODKA` was sold, which vendor has the most transactions? Report the total number of transactions, the number of bottles sold, and the total amount of revenue from these transactions. Rename the columns to make sure there isn't any confusion as to what they mean!

In [165]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT vendor, 
    COUNT(vendor) as transactions, 
    SUM(bottle_qty) as bottles_sold, 
    SUM(total) as revenue
FROM sales
WHERE category_name = 'IMPORTED VODKA'
GROUP BY vendor
ORDER BY revenue desc
"""

pd.read_sql_query(sql, engine)

vendor  transactions  bottles_sold  \
0       Pernod Ricard USA/Austin Nichols         42338        403450   
1                    Bacardi U.S.A. Inc.         26441        236047   
2                        Diageo Americas         15725        152038   
3        Constellation Wine Company Inc.         12565        116053   
4                        Jim Beam Brands          5850        104119   
5                            Stoli Group          4443         35513   
6                   Bacardi U.S.A., Inc.          3344         30239   
7                 Moet Hennessy USA Inc.          2153         10966   
8                   Imperial Brands Inc.          2799         21033   
9       Constellation Wine Company, Inc.          1672         12956   
10                               Proximo          1628          9431   
11                        Luxco-St Louis          1128          8933   
12                               MHW Ltd          1159          4493   
13                   Wilson Daniels Ltd.           815          2396   
14            Russian Standard Vodka USA           415          2421   
15                 Imperial Brands, Inc.           426          3054   
16                                 LUXCO           140          2374   
17               Moet Hennessy USA, Inc.           256          1032   
18                      Quadro Group LLC           289          1054   
19              Brown-Forman Corporation           263          1730   
20           WILLIAM GRANT AND SONS INC.           232          1119   
21                      Klin Spirits LLC           218           784   
22  Shaw Ross International Importers LL            10           516   
23      Western Spirits Beverage Co. LLC            18           484   
24                 A Hardy / U.S.A. Ltd.            33           320   
25           W. J. Deutsch And Sons Ltd.            24           768   
26                     Jinro America Inc            21          1056   
27                                  None             0           368   
28           Russian Standard Vodka, USA            71           360   
29          WILLIAM GRANT AND SONS, INC.            56           252   
30                     Wine Intelligence             3           360   
31                             TY KU LLC            17            87   
32            The Patron Spirits Company             4            54   
33          Park Street Imports LLC Voli            12            84   
34                      Sazerac Co. Inc.            10            60   
35              Firestarter Spirits Inc.            20            54   
36      Stanley Stawski Distributing Co.             2            24   
37                 Ti Beverage Group Ltd             2            24   
38        Charles Jacquin Et. Cie., Inc.             2            12   
39                     Klin Spirits, LLC             5            13   
40                     Phenix Brands LLC             2            12   
41             Adamba Imports Int'l Inc.             1            12   
42                A Hardy / U.S.A., Ltd.             1             5   

       revenue  
0   7880365.88  
1   5771242.07  
2   3346425.42  
3   1895005.51  
4   1807097.23  
5    832071.19  
6    765347.31  
7    323126.54  
8    265317.79  
9    200283.64  
10   178212.03  
11   116914.19  
12   109190.48  
13    84246.72  
14    40361.58  
15    37950.20  
16    32425.32  
17    30783.53  
18    25928.40  
19    23337.88  
20    20981.25  
21    15483.84  
22    13731.84  
23     9453.60  
24     9120.00  
25     8593.92  
26     7319.04  
27     6009.44  
28     5400.00  
29     4725.00  
30     3106.80  
31     2429.85  
32     1827.36  
33     1611.96  
34     1140.00  
35      994.14  
36      498.00  
37      462.00  
38      294.84  
39      260.13  
40      233.28  
41      107.88  
42      107.55

**Prompt 5 Answer**: Pernod Ricard USA/Austin Nichols, 42338 transactions,	403450 bottles sold, $7,880,365.88 in revenue.

### Prompt 6

The reason we sell things is to make money. In sales, the term **markup** means the amount of extra money charged for a product over the cost to make that product. (You can think of the markup as the profit for that product.)

In Iowa, the law states the minimum price at which liquor may be sold.

Create a column called `Markup` that shows the difference between the `btl_price` and the `state_btl_cost`. Take a screenshot and paste it into your notebook.

> If you haven't subtracted columns together before, [this StackOverflow question](https://stackoverflow.com/questions/7536996/multiplying-two-columns-in-sql-server/7537059) may be helpful.

In [166]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT btl_price - state_btl_cost AS markup
FROM sales

"""

pd.read_sql_query(sql, engine)

markup
0         $5.75
1         $5.75
2         $5.75
3         $8.58
4         $2.50
...         ...
3049908   $5.75
3049909  $10.00
3049910   $5.75
3049911   $5.75
3049912  $10.00

[3049913 rows x 1 columns]

**Prompt 6 Answer**: 

![Image](markup.png)

### Prompt 7

Calculate the "Percentage Markup" by dividing markup by the state bottle cost. (Rather than the total profit per bottle, this will tell you how much stores are increasing their price over the state minimum.)

After calculating the "Percentage Markup," calculate the average percentage markup per **store**. Which store has the largest average markup, and what is that average markup?

> Note: To make this an easier question, you may take the simple average of percentage markups across all transactions for a store. You do **not** need to weight the average markup by the number of bottles purchased per transaction. If you want to practice your SQL skills, though, consider this a bonus challenge!

In [167]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT 
st.store AS store_iD, 
st.name AS store_name,
st.store_address AS store_address,
AVG((btl_price - state_btl_cost)/state_btl_cost) AS avg_pct_markup
FROM sales AS sa
INNER JOIN stores AS st ON st.store = sa.store
GROUP BY st.store
ORDER BY avg_pct_markup DESC
"""

pd.read_sql_query(sql, engine)

store_id                            store_name  \
0         4013                Wal-Mart 0841 / Tipton   
1         4024            Wal-Mart 1546 / Iowa Falls   
2         4266            Wal-Mart 1683 / Shenandoah   
3         3833              Wal-Mart 3394 / Atlantic   
4         3660             Wal-Mart 2935 / Knoxville   
...        ...                                   ...   
1316      9018           Paradise Distilling Company   
1317      4784  Hawkeye Convenience Store Wiley Blvd   
1318      4494              SELECT MART / SIOUX CITY   
1319      4534                           D & B Foods   
1320      4693  Hawkeye Convenience Stores / 16th St   

                                          store_address  avg_pct_markup  
0     1126 Highway 38 North\nTipton, IA 527720000\n(...        0.525471  
1     840 S Oak\nIowa Falls, IA 501260000\n(42.50295...        0.521499  
2     705 S Fremont\nShenandoah, IA 516010000\n(40.7...        0.518903  
3     1905 East 7th St\nAtlantic, IA 500220000\n(41....        0.515094  
4     814 W Bell Ave\nKnoxville, IA 501380000\n(41.3...        0.514591  
...                                                 ...             ...  
1316  245, Railroad Ave\nDubuque, IA 52001\n(42.4873...        0.500000  
1317  2330 Wiley Blvd Sw\nCedar Rapids, IA 52404\n(4...        0.499994  
1318  800 Hamilton Blvd\nSioux City, IA 51103\n(42.5...        0.499909  
1319  141 W Gillet St\nPreston, IA 52069\n(42.050293...        0.499781  
1320  250 16th Ave Sw\nCedar Rapids, IA 52404\n(41.9...        0.499214  

[1321 rows x 4 columns]

**Prompt 7 Answer**: The Tipton Wal-Mart, avg markup: ~53%

### Prompt 8

Building off of your last query, retrieve the five stores with the highest average markup percentage. Along with the store IDs and average markup percentage, please return the minimum and maximum markup percentage for those stores, as well as the names and addresses of those store.

What are the names of the five stores?

In [168]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT 
st.store AS store_iD, 
st.name AS store_name,
st.store_address AS store_address,
AVG((btl_price - state_btl_cost)/state_btl_cost) AS avg_pct_markup,
MAX((btl_price - state_btl_cost) / state_btl_cost) AS max_pct_markup,
MIN((btl_price - state_btl_cost) / state_btl_cost) AS min_pct_markup
FROM sales AS sa
INNER JOIN stores AS st ON st.store = sa.store
GROUP BY st.store, st.store_address
ORDER BY avg_pct_markup DESC
LIMIT 5
"""

pd.read_sql_query(sql, engine)

store_id                  store_name  \
0      4013      Wal-Mart 0841 / Tipton   
1      4024  Wal-Mart 1546 / Iowa Falls   
2      4266  Wal-Mart 1683 / Shenandoah   
3      3833    Wal-Mart 3394 / Atlantic   
4      3660   Wal-Mart 2935 / Knoxville   

                                       store_address  avg_pct_markup  \
0  1126 Highway 38 North\nTipton, IA 527720000\n(...        0.525471   
1  840 S Oak\nIowa Falls, IA 501260000\n(42.50295...        0.521499   
2  705 S Fremont\nShenandoah, IA 516010000\n(40.7...        0.518903   
3  1905 East 7th St\nAtlantic, IA 500220000\n(41....        0.515094   
4  814 W Bell Ave\nKnoxville, IA 501380000\n(41.3...        0.514591   

   max_pct_markup  min_pct_markup  
0        6.708763        0.498571  
1        6.708763        0.498462  
2        6.708763        0.498462  
3        6.708763        0.498462  
4        6.708763        0.498462

**Prompt 8 Answer**: WAL-MART STRIKES AGAIN

### Prompt 9

Which nine counties sell the highest number of bottles of liquor per capita (per person)? Calculate the average bottle size of the bottles of liquor sold by these counties rounded to the nearest first decimal place. Of the nine counties selling the highest number of bottles of liquor per capita, which county sells (on average) the largest bottle size?

**_BONUS FUN_**: This question can (also) be solved with a SQL subquery - that is, a query that queries a query! If you'd like to give this a try, you can read more about it [here](https://www.dofactory.com/sql/subquery). Otherwise, keep an eye out for the solution key!

Got tired of Scrolling Up:
<img src="./images/schema.png" alt="schema" width="750"/>

In [169]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT c.county, 
       c.population,
       (SUM(s.bottle_qty)/c.population) AS btls_sold_percap,
       ROUND(AVG(p.bottle_size),1) AS avg_btl_size
FROM sales AS s
INNER JOIN counties AS c ON c.county = s.county
INNER JOIN products AS p ON p.item_no = s.item
GROUP BY c.county
ORDER BY btls_sold_percap DESC
LIMIT 9
"""

pd.read_sql_query(sql, engine)

county  population  btls_sold_percap  avg_btl_size
0      Dickinson       16667                22         974.1
1     Black Hawk      131090                15         862.9
2           Polk      430640                15         892.6
3    Cerro Gordo       44151                14         986.4
4          Scott      165224                13         871.6
5        Johnson      130882                13         895.4
6           Linn      211226                12         886.3
7        Kossuth       15543                11        1041.0
8  Pottawattamie       93158                11         893.7

**Prompt 9 Answer**: Kossuth County

### Prompt 10 (advanced but required!)

In the `sales` table, three columns are `btl_price`, `bottle_qty`, and `total`. The `total` for a transaction _should be_ the product of `btl_price` and `bottle_qty`. How many transactions have a value of `total` that is not equal to `btl_price` time `bottle_qty`?

> There will be a type error as well! We cannot compare type `money` to type `real`. We have to convert `total` to type `money` so that we can compare the two directly. [This link](https://docs.microsoft.com/en-us/sql/t-sql/functions/cast-and-convert-transact-sql?view=sql-server-ver15#syntax) may be helpful to you!

<img src="./images/schema.png" alt="schema" width="750"/>

In [215]:
# Use a SQL query to find the answer to the above prompt.

sql = """

SELECT *
           
FROM sales AS s

WHERE CAST(s.total AS money) != CAST((s.bottle_qty * s.btl_price) AS money)
"""

pd.read_sql_query(sql, engine)

Empty DataFrame
Columns: [date, convenience_store, store, county_number, county, category, category_name, vendor_no, vendor, item, description, pack, liter_size, state_btl_cost, btl_price, bottle_qty, total]
Index: []

**Prompt 10 Answer**: It seems no records are unequal to the expected price? 